File: CodeForces/Contest/Round #673 (Div 1)/C673.cpp
#include <bits/stdc++.h>

using namespace std;


int main(){
   ios::sync_with_stdio(false);
   cin.tie(0);
   
   int n;
   cin >> n;
   
   vector<int> array;
   for(int k = 0; k < n; k++){
      int x;
      cin >> x;
      array.push_back(x);
   }
   
   long long numinversions = 0LL;
   int num = 0;
   
   for(int i = 30; i >= 1; i--){
      unordered_map<int,vector<int>> hmap;
      int ishifti1 = 1 << (i-1);
      for(int k = 0; k < n; k++){
         int shift = array[k] >> i;
            
         if((array[k] & (ishifti1)) > 0){
            hmap[shift].push_back(1);
         } 
         else {
            hmap[shift].push_back(0);
         }
            
      }
      
      long long flip = 0LL;
      long long noflip = 0LL;
      for(auto x : hmap){
         long long num0 = 0L;
         long long num1 = 0L;
         for(int v : x.second){
            if(v == 1){
               flip += num0;
               num1++;
            } 
            else {
               noflip += num1;
               num0++;
            }
         }
      }
      
      if(flip < noflip){
         num += ishifti1;
      }
         
      numinversions += min(flip,noflip);
   }
   
   cout << numinversions << " " << num << endl;
   
   
   
   
   return 0;
}

File: CodeForces/Contest/Round #673 (Div 1)/C673.java
//make sure to make new file!
import java.io.*;
import java.util.*;

public class C673{
   
   public static void main(String[] args)throws IOException{
      BufferedReader f = new BufferedReader(new InputStreamReader(System.in));
      PrintWriter out = new PrintWriter(System.out);
      
      int n = Integer.parseInt(f.readLine());
      
      StringTokenizer st = new StringTokenizer(f.readLine());
      
      int[] array = new int[n];
      for(int k = 0; k < n; k++){
         array[k] = Integer.parseInt(st.nextToken());
      }
      
      long numinversions = 0L;
      int num = 0;
      HashMap<Integer,ArrayList<Integer>> hmap = new HashMap<Integer,ArrayList<Integer>>();
      for(int i = 30; i >= 1; i--){
         //see if ith decimal should be flipped
         //divide every number by 1 << i, so >> i
         //group into arraylist
         
         int ishifti1 = 1 << (i-1);
         
         
         for(int k = 0; k < n; k++){
            int shift = array[k] >> i;
            if(!hmap.containsKey(shift)){
               hmap.put(shift,new ArrayList<Integer>());
            }
            
            if((array[k] & (ishifti1)) > 0){
               hmap.get(shift).add(1);
            } else {
               hmap.get(shift).add(0);
            }
            
         }
         
         long flip = 0L;
         long noflip = 0L;
         for(int x : hmap.keySet()){
            long num0 = 0L;
            long num1 = 0L;
            for(int v : hmap.get(x)){
               if(v == 1){
                  flip += num0;
                  num1++;
               } else {
                  noflip += num1;
                  num0++;
               }
            }
         }
         
         if(flip < noflip){
            num += ishifti1;
         }
         
         numinversions += Math.min(flip,noflip);
         hmap.clear();
      }
      
      out.println(numinversions + " " + num);
   
      
      
      
      
      
      
      
      out.close();
   }
   
      
}